osutil/resolve_path.go

```diff
@@ -0,0 +1,117 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package osutil
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+func dumbJoin(a, b string) string {
+	if strings.HasSuffix(a, "/") {
+		return a + b
+	} else {
+		return a + "/" + b
+	}
+}
+
+func resolvePathInSysrootRec(sysroot, path string, validInAnySysroot bool, symlinkRecursion int) (string, error) {
+	if path == "" || path == "/" {
+		// Relative paths are taken from sysroot
+		return "/", nil
+	}
+
+	if strings.HasSuffix(path, "/") {
+		path = path[:len(path)-1]
+	}
+
+	dir, file := filepath.Split(path)
+	resolvedDir, err := resolvePathInSysrootRec(sysroot, dir, validInAnySysroot, symlinkRecursion)
+	if err != nil {
+		return "", err
+	}
+	if file == "" {
+		return resolvedDir, nil
+	}
+	if file == "." {
+		return resolvedDir, nil
+	}
+	if file == ".." {
+		if validInAnySysroot && (resolvedDir == "/") {
+			return "", fmt.Errorf("invalid escaping path")
+		}
+		upperDir, _ := filepath.Split(resolvedDir)
+		return upperDir, nil
+	}
+
+	fileInResolvedDir := dumbJoin(resolvedDir, file)
+
+	realPath := dumbJoin(sysroot, fileInResolvedDir)
+	st, err := os.Lstat(realPath)
+	if err != nil {
+		return "", err
+	}
+
+	if st.Mode()&os.ModeSymlink != 0 {
+		if symlinkRecursion < 0 {
+			return "", fmt.Errorf("Maximum recursion reached when reading symlinks")
+		}
+		target, err := os.Readlink(realPath)
+		if err != nil {
+			return "", err
+		}
+		if filepath.IsAbs(target) {
+			if validInAnySysroot {
+				return "", fmt.Errorf("Invalid absolute symlink")
+			}
+			return resolvePathInSysrootRec(sysroot, target, validInAnySysroot, symlinkRecursion-1)
+		} else {
+			return resolvePathInSysrootRec(sysroot, dumbJoin(resolvedDir, target), validInAnySysroot, symlinkRecursion-1)
+		}
+	}
+
+	return fileInResolvedDir, nil
+}
+
+// Resolve a path within a sysroot
+//
+// In a sysroot, abolute symlinks should be relative to the sysroot
+// rather than `/`. Also paths with multiple `..` that would escape
+// the sysroot should not do so.
```

Comment:
This two cases should come with examples of the results I think in the doc comment for clarity.

Reply:
I have added 3 examples. I have used 3 layers of indirection to make it more obvious this is not just a dumb `os.Readlink`.